pkg/appgateserver/docs/server.md

```diff
@@ -0,0 +1,278 @@
+ones too. These are seen as the same at the protocol level_
+
+In the case there is an internal server error, with the `AppGateServer` itself,
+the payload must be partially unmarshalled to retrieve specific information.
+This information is required for the error to be returned not only in the
+correct format, but also to contain specified fields.
+
+For example a JSON-RPC request will contain the following fields:
+
+```json
+{
+    "id": 1,                        // id of the request
+    "jsonrpc": "2.0"                // JSON-RPC version number
+    "method": "eth_getBlockNumber", // the method to be executed
+    "params": {},                   // either a list or map of params
+}
+```
+
+When returning an internal server error the response must contain the fields:
+
+```json
+{
+    "id": 1                // this must be the same as from the request
+    "jsonrpc": "2.0"       // this must be the same as from the request
+    "error": {
+        "code":    -32000, // the error code of an internal error
+        "message": "",     // the error message
+        "data":    null,   // a nil data field
+    },
+}
+```
+
+In order to extract the relevent fields from the serialised payload we must
+partially unmarshal it, extracting only the desired fields. The same logic
+is used for other types of symmetric protocols, extracting different fields for
+each payload type.
+
+#### Asymmetric Protocols
+
+_NOTE: Asymmetric Protocols are currently unsupported as of writing this
+document_
+
+## Sessions
+
+The `AppGateServer` will handle the interaction with the chain to retrieve the
+current session for the application whose request it is handling. This enables
+the application to simply send their requests without having to manage their
+sessions or keep track of them.
+
+The `AppGateServer` will query the chain for the current session ID for the
+given application and service pair.
+
+See [session.go](../session.go) for more details on the implementation of
+session retrieval.
+
+## Endpoint Selection
+
+Endpoints for the suppliers in a session are currently chosen based on the
+first suitable supplier for the service from within the session.
+
+_NOTE: The endpoint selection algorithm is planned to factor in other factors
+such as Quality of Service and move away from the greedy selection
+implementation_
+
+See [endpoint-selector.go](../endpoint_selector.go) for more details on the
+implementation of supplier endpoint selection.
+
+## Signatures
+
+The `AppGateServer` signs all requests (as both applications and gateways) using
+the ring of the application sending the request. They sign the ring using the
+private key associated with the key used when starting the server.
+
+When a response is received from the supplier, it is verified using the public
+key of the supplier who made the request.
+
+### Rings
+
+Rings are created using the application's on-chain state. This includes a list
```

Comment:
```suggestion
Rings are created using the application's on-chain state. In gateway mode, this includes a list
```

Reply:
This will always include the gateways, it depends on whether the application has delegated or not, but if it has and it starts its own appgateserver to self sign it will still have those in its ring